### PR TITLE
[go_router] Adds `void replace()` and `replaceNamed` to `GoRouterDelegate`, `GoRouter` and `GoRouterHelper`

### PR DESCRIPTION
**Pull request by ValentinVignal**
This PR Adds `void replace()` and `replaceNamed` to `GoRouterDelegate`, `GoRouter` and `GoRouterHelper`.

It should fix https://github.com/flutter/flutter/issues/106402

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router/lib/go_router.dart [hunk 1]
```diff
@@ -70,6 +70,26 @@ extension GoRouterHelper on BuildContext {
         extra: extra,
       );
 
+  /// Replaces the current location with the given one w/ optional query
```
Comment (chunhtai) [inline packages/go_router/lib/go_router.dart:73]:
```suggestion
  /// Replaces the top-most page of page stack with the given URL location w/ optional query
```

File: packages/go_router/lib/go_router.dart [hunk 2]
```diff
@@ -70,6 +70,26 @@ extension GoRouterHelper on BuildContext {
...
+  /// parameters, e.g. `/family/f2/person/p1?color=blue
```
Comment (chunhtai) [inline packages/go_router/lib/go_router.dart:74]: maybe also mention go/push use
```
See also:

 * [go], which ....
 * [push], which ...
```

File: packages/go_router/lib/go_router.dart [hunk 3]
```diff
@@ -70,6 +70,26 @@ extension GoRouterHelper on BuildContext {
...
+  void replace(String location, {Object? extra}) =>
+      GoRouter.of(this).replace(location, extra: extra);
+
+  /// Replaces the current location with the named route w/ optional parameters,
```
Comment (chunhtai) [inline packages/go_router/lib/go_router.dart:78]:
```suggestion
  /// Replaces the top-most page of page stack with the named route w/ optional parameters,
```

**Review by chunhtai — COMMENTED** [review]
code looks good, left some documentation nits

**Comment by ValentinVignal** [comment]
Thank you @chunhtai, I have applied the changes in https://github.com/flutter/packages/pull/2306/commits/33f74249df68226cea85fe6cb5ce00fe396074f3

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by chunhtai** [comment]
there is a merge conflict, can you resolve them before this pr can merge?

**Comment by ValentinVignal** [comment]
> there is a merge conflict, can you resolve them before this pr can merge?

I fixed the conflicts in https://github.com/flutter/packages/pull/2306/commits/a9019f920622b23f72202b78e34bb3cb7ffbb70b

File: packages/go_router/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,6 @@
+## 4.2.0
+
+- Adds `void replace()` and `replaceNamed` to `GoRouterDelegate`, `GoRouter` and `GoRouterHelper`.
```
Comment (chunhtai) [inline packages/go_router/CHANGELOG.md:3]: nit: add an empty line between two versions

File: packages/go_router/test/go_router_delegate_test.dart [hunk 5]
```diff
@@ -78,6 +78,52 @@ void main() {
     );
   });
 
+  group('replace', () {
```
Comment (johnpryan) [inline packages/go_router/test/go_router_delegate_test.dart:81]: consider adding a test for `replaceNamed` too?

**Review by johnpryan — APPROVED** [review]

File: packages/go_router/test/go_router_delegate_test.dart (hunk 5 above)
Comment (ValentinVignal) [reply packages/go_router/test/go_router_delegate_test.dart:81]: I added a test for `replaceNamed` in https://github.com/flutter/packages/pull/2306/commits/44e35ef220cb98e5b34d89bcd2d7c7c3083b3426 Tell me what you think about it

**Pull request merged**